### PR TITLE
HardwareDevices: Smart data for NVMe drives

### PR DESCRIPTION
**Pull request by ge0rdi**
Code based on:
https://learn.microsoft.com/en-us/windows/win32/fileio/working-with-nvme-devices#example-nvme-get-log-pages-query

Description of SMART NVMe attributes:
https://en.wikipedia.org/wiki/Self-Monitoring,_Analysis_and_Reporting_Technology#Known_NVMe_S.M.A.R.T._attributes

How it looks like:
image

### CODE REVIEW

**Review by jxy-s — APPROVED** [review]

File: plugins/HardwareDevices/diskdetails.c [hunk 1]
```diff
@@ -107,6 +107,23 @@ VOID DiskDeviceAddListViewItems(
     }
 }
 
+VOID AddNvmeSmartEntry(HWND ListViewHandle, LPCWSTR Name, PPH_STRING Value)
```
Comment (jxy-s) [inline plugins/HardwareDevices/diskdetails.c:110]:
```suggestion
VOID AddNvmeSmartEntry(
    _In_ HWND ListViewHandle, 
    _In_ LPCWSTR Name, 
    _In_ PPH_STRING Value
    )
```

File: plugins/HardwareDevices/diskdetails.c [hunk 2]
```diff
@@ -107,6 +107,23 @@ VOID DiskDeviceAddListViewItems(
...
+{
+    INT lvItemIndex = PhAddListViewItem(
+        ListViewHandle,
+        MAXINT,
+        Name,
+        0
+    );
```
Comment (jxy-s) [inline plugins/HardwareDevices/diskdetails.c:117]:
```suggestion
        );
```

File: plugins/HardwareDevices/diskdetails.c [hunk 3]
```diff
@@ -107,6 +107,23 @@ VOID DiskDeviceAddListViewItems(
...
+
+    PhSetListViewSubItem(
+        ListViewHandle,
+        lvItemIndex,
+        1,
+        PhGetString(Value)
+    );
```
Comment (jxy-s) [inline plugins/HardwareDevices/diskdetails.c:124]:
```suggestion
        );
```

File: plugins/HardwareDevices/storage.c [hunk 4]
```diff
@@ -563,6 +563,66 @@ NTSTATUS DiskDriveQueryImminentFailure(
...
+
+    status = PhDeviceIoControlFile(
+        DeviceHandle,
+        IOCTL_STORAGE_QUERY_PROPERTY,
+        buffer,
+        sizeof(buffer),
+        buffer,
+        sizeof(buffer),
+        &returnedLength
+    );
```
Comment (jxy-s) [inline plugins/HardwareDevices/storage.c:602]:
```suggestion
        );
```

**Review by jxy-s — CHANGES_REQUESTED** [review]
Should probably hide the irrelevant columns when displaying the NVME SMART data (`Best`, `Raw`, `Raw (Hex)`,...)?

**Comment by ge0rdi** [comment]
> Should probably hide the irrelevant columns when displaying the NVME SMART data (Best, Raw, Raw (Hex),...)?

That would be great.
Unfortunately, hiding/deleting columns seems to interfere with saving column widths.
So in case if you have mix of NVMe and non-NVMe drives, column widths for `Best`/`Raw`/etc  won't get saved (if you have displayed Smart properties for NVMe drive).

**Review by jxy-s — APPROVED** [review]

**Pull request merged**